sympy/solvers/solveset.py

```diff
@@ -918,7 +923,13 @@ def solveset(f, symbol=None, domain=S.Complexes):
             result = ConditionSet(symbol, f, domain)
         return result
 
-    return _solveset(f, symbol, domain, _check=True)
+    if decompogen(f, symbol)[0] == f:
+        return _solveset(f, symbol, domain, _check=True)
+    else:
+        try:
+            return solve_decomposition(f, symbol, domain)
+        except:
+            return _solveset(f, symbol, domain, _check=True)
```

Comment:
> Why there is try block?

Several cases that are not solved by `solve_decomposition` fail. Mainly `TypeError: 'ConditionSet' object is not iterable` exception is returned.
`solveset(abs(x) - n, x, S.Reals)`
`solveset((a*x + b)*(exp(x) - 3), x, S.Reals)`, and almost every such case